src/apps/chifra/pkg/cache/read.go

```diff
@@ -0,0 +1,813 @@
+		return
+	}
+
+	err = read(&tx.Nonce)
+	if err != nil {
+		return
+	}
+
+	err = readTimestamp(read, &tx.Timestamp)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &tx.From)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &tx.To)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &tx.Value)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &tx.ExtraValue1)
+	if err != nil {
+		return
+	}
+
+	err = readBigUint(reader, &tx.ExtraValue2)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.Gas)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.GasPrice)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.MaxFeePerGas)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.MaxPriorityFeePerGas)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &tx.Input)
+	if err != nil {
+		return
+	}
+
+	err = readUintAsBool(read, &tx.IsError)
+	if err != nil {
+		return
+	}
+
+	err = readUintAsBool(read, &tx.HasToken)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.Cachebits)
+	if err != nil {
+		return
+	}
+
+	err = read(&tx.Reserved2)
+	if err != nil {
+		return
+	}
+
+	receipt, err := ReadReceipt(reader)
+	if err != nil {
+		return
+	}
+	tx.Receipt = receipt
+
+	err = readFromArray(reader, &tx.Traces, ReadTrace)
+	if err != nil {
+		return
+	}
+
+	articulatedTx, err := ReadFunction(reader)
+	if err != nil {
+		return
+	}
+	tx.ArticulatedTx = articulatedTx
+
+	return
+}
+
+func ReadReceipt(reader *bufio.Reader) (receipt *types.SimpleReceipt, err error) {
+	receipt = &types.SimpleReceipt{}
+	read := createReadFn(reader)
+
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &receipt.ContractAddress)
+	if err != nil {
+		return
+	}
+
+	err = read(&receipt.GasUsed)
+	if err != nil {
+		return
+	}
+
+	err = read(&receipt.EffectiveGasPrice)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &receipt.Logs, ReadLog)
+	if err != nil {
+		return
+	}
+
+	err = read(&receipt.Status)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func ReadLog(reader *bufio.Reader) (log *types.SimpleLog, err error) {
+	log = &types.SimpleLog{}
+	read := createReadFn(reader)
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &log.Address)
+	if err != nil {
+		return
+	}
+
+	err = read(&log.LogIndex)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &log.Topics, makeArrayItemRead(reader, readHash))
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &log.Data)
+	if err != nil {
+		return
+	}
+
+	articulatedLog, err := ReadFunction(reader)
+	if err != nil {
+		return
+	}
+	log.ArticulatedLog = articulatedLog
+
+	return
+}
+
+func ReadFunction(reader *bufio.Reader) (function *types.SimpleFunction, err error) {
```

Comment:
Eventually, we'll move this into the auto-generated type specific file. Please add a TODO, but don't move this yet.